main_code/Case_1/warehouse_reminder.py

```diff
@@ -0,0 +1,282 @@
+        self.id = part_id
+        self.date = self.convert_to_date_obj(date)
+        self.amount = amount
+        self.open = open
+        self.email = contact
+        self.from_email = from_email
+        self.server = server
+
+    def __repr__(self):
+        return('ID: {id}, Due: {date}, Recieved: {amount}, '
+              'In-Shipping: {open}, Contact: {email}, Send From: {from_email}'.format(
+               id=self.id, date=self.date, amount=self.amount, open=self.open, email=self.email,
+               from_email=self.from_email
+        ))
+
+    def check_if_action_required(self) -> bool:
+        """
+        Determines if action is required based on date and value of open and amount
+        :return: bool
+        """
+
+        current_time = datetime.datetime.today()
+
+        # we store our date as an object called date, but datetime uses objects called datetime
+        # so we use .combine our date object with the minimal time interval to get the right
+        # kind of object
+        if datetime.datetime.combine(self.date,
+                                     datetime.datetime.min.time()) - datetime.timedelta(days=5) <= current_time:
+            return True
+        return False
+
+    def send_email(self) -> None:
+        """
+        Sends email with the part ID, how many ordered, how many still accepting as reminder
+        :return: None
+        """
+
+        # just like earlier we use this to convert date into datetime
+        days_between_deadline_and_now = datetime.datetime.combine(
+            self.date, datetime.datetime.min.time()) - datetime.datetime.today()
+
+        static_email_text = '''
+        Hello,
+        \tThis is an automated email reminder about ordering part {part_id}.  At this time only {amount}
+        have been recieved and {open} are currently on route.  As of now there are less than {days} left
+        before the deadline on {date} has passed.
+        '''.format(open=self.open, part_id=self.id, amount=self.amount, date=str(self.date),
+                   days=int(days_between_deadline_and_now.days))
+
+        msg = MIMEMultipart()
+        msg['Subject'] = 'AUTOMATED Part Order Reminder'
+        msg['From'] = self.from_email
+        msg['To'] = self.email
+        # msg.attach(static_email_text)
+        body = MIMEText(static_email_text)
+        msg.attach(body)
+
+        self.server.sendmail(self.from_email, self.email, msg.as_string())
+
+    @staticmethod
+    def convert_to_date_obj(date_str: Any) -> datetime.date:
+        """
+        attempts to convert date string into a usable date object
+        :param date_str: date string grabbed
+        :return: datetime.date object
+        """
+
+        if type(date_str) is datetime.datetime:
+            return date_str
+        try:
+            dt = parser.parse(date_str)
+        except Exception as e:
+            print('Unable to parse date string: %s' % date_str)
```

Comment:
What would be the situation that it can not convert date?

Reply:
If it's a bad string or really ugly or blank. IE for example

Janary 18 2019

It might be able to handle this but it might not, and it's definitely not capable of handing empty strings so it's safer to try it and throw an exception in processing than to just hope.